dom.go

```diff
@@ -347,89 +465,152 @@ func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts
+				// nodes), rather than replacing the previous child at this
+				// position.
+				if insertBeforeKeyedNode != nil {
+					// Insert before the next sibling, if we have one.
+					h.insertBefore(insertBeforeKeyedNode, nextChildRender, prevChildRender)
+					continue
+				}
+				h.insertBefore(h.insertBeforeNode, nextChildRender, prevChildRender)
+				continue
+			}
 			replaceNode(nextChildRender.node, prevChildRender.node)
 		case nextChildRender == nil && prevChildRender != nil:
+			// If we're removing the current insert target, use the next
+			// sibling, if any.
+			if h.insertBeforeNode != nil && h.insertBeforeNode == prevChildRender.node {
+				h.insertBeforeNode = h.insertBeforeNode.Get("nextSibling")
+			}
 			unmount(prevChild)
 			removeNode(prevChildRender.node)
 		case nextChildRender != nil && prevChildRender == nil:
 			if m, ok := nextChild.(Mounter); ok {
 				pendingMounts = append(pendingMounts, m)
 			}
-			if insertBefore != nil {
-				h.node.Call("insertBefore", nextChildRender.node, insertBefore.node)
+			if insertBeforeKeyedNode != nil {
+				// Insert before the next keyed sibling, if we have one.
+				h.insertBefore(insertBeforeKeyedNode, nextChildRender, prevChildRender)
 				continue
 			}
-			h.node.Call("appendChild", nextChildRender.node)
+			h.insertBefore(h.insertBeforeNode, nextChildRender, prevChildRender)
 		default:
 			panic("vecty: internal error (unexpected switch state)")
 		}
 	}
-	h.removeChildren(prev)
+
+	// If dealing with keyed siblings, remove children by key rather than
+	// position.
+	if keyed {
+		// Do not remove dangling keys yet if we're inside a nested childList.
+		if h.nestedKeys {
+			return pendingMounts
+		}
+		prevChildren := make([]ComponentOrHTML, len(prev.keyMap))
+		i := 0
+		for _, c := range prev.keyMap {
+			prevChildren[i] = c
+			i++
+		}
+		h.removeChildren(prevChildren)
+		return pendingMounts
+	}
+
+	if len(prev.children) > len(h.children) {
+		// Every previous child that h.children does not have in common.
+		h.removeChildren(prev.children[len(h.children):])
+	}
 	return pendingMounts
 }
 
 // removeChildren removes child elements from the previous render pass that no
 // longer exist on the current HTML children.
-func (h *HTML) removeChildren(prev *HTML) {
-	if prev == nil {
-		return
-	}
-	// Every previous child that h.children does not have in common.
-	for i := len(h.children); i < len(prev.children); i++ {
-		prevChild := prev.children[i]
-		if prevChildList, ok := prevChild.(List); ok {
+func (h *HTML) removeChildren(prevChildren []ComponentOrHTML) {
+	for _, prevChild := range prevChildren {
+		if prevChildList, ok := prevChild.(childList); ok {
 			// Previous child was a list, so remove all DOM nodes in it.
-			prevChildList.remove(h.node)
+			prevChildList.remove(h, h.insertBeforeNode)
+			if h.insertBeforeNode != nil {
+				h.insertBeforeNode = prevChildList.html.insertBeforeNode
+			}
```

Comment:
I don't understand what this if block is doing. Please document.

Reply:
The first line is documented at [diff](https://github.com/gopherjs/vecty/pull/147/files/1371c31a8088ea66eb45d93dba67c0f75d4aabb0#diff-cf11d578486f49277bfe20d59ea10cecR419), the last three are the same as [diff](https://github.com/gopherjs/vecty/pull/147/files/1371c31a8088ea66eb45d93dba67c0f75d4aabb0#diff-cf11d578486f49277bfe20d59ea10cecR408), I'll copy the comment.